import numpy as np
import pandas as pd

To rank the rows of Pandas DataFrame we can use the DataFrame.rank() method which returns a rank of every respective index of a series passed. The rank is returned on the basis of position after sorting.

Example #1 :

Here we will create a DataFrame of movies and rank them based on their ratings.

In [5]:
# Define the dictionary for converting to dataframe  
movies = {'Name': ['The Godfather', 'Bird Box', 'Fight Club'], 
         'Year': ['1972', '2018', '1999'], 
         'Rating': ['9.2', '6.8', '8.8']} 
  
df = pd.DataFrame(movies) 
print(df) 

            Name  Year Rating
0  The Godfather  1972    9.2
1       Bird Box  2018    6.8
2     Fight Club  1999    8.8


In [27]:
# Create a column Rating_Rank which contains 
# the rank of each movie based on rating 
df['Rating_Rank'] = df['Rating'].rank(ascending=True)
# Set the index to newly created column, Rating_Rank 
df = df.set_index('Rating_Rank')
print(df)

                      Name  Year Rating
Rating_Rank                            
1.0               Bird Box  2018    6.8
2.0             Fight Club  1999    8.8
3.0          The Godfather  1972    9.2


In [15]:
# Sort the dataFrame based on the index 
df = df.sort_index()
df

,Name,Year,Rating
Rating_Rank,,,
1.0,Bird Box,2018,6.8
2.0,Fight Club,1999,8.8
3.0,The Godfather,1972,9.2


Example #2

Let’s take an example of marks scored by 4 students. We will rank the students based on the highest mark they have scored.

In [9]:
# Create a dictionary with student details 
student_details = {'Name':['Raj', 'Raj', 'Raj', 'Aravind', 'Aravind', 'Aravind', 
                             'John', 'John', 'John', 'Arjun', 'Arjun', 'Arjun'], 
                   'Subject':['Maths', 'Physics', 'Chemistry', 'Maths', 'Physics', 
                             'Chemistry', 'Maths', 'Physics', 'Chemistry', 'Maths', 
                                                           'Physics', 'Chemistry'], 
                   'Marks':[80, 90, 75, 60, 40, 60, 80, 55, 100, 90, 75, 70] 
               } 
  
# Convert dictionary to a DataFrame 
df = pd.DataFrame(student_details) 
print(df) 

       Name    Subject  Marks
0       Raj      Maths     80
1       Raj    Physics     90
2       Raj  Chemistry     75
3   Aravind      Maths     60
4   Aravind    Physics     40
5   Aravind  Chemistry     60
6      John      Maths     80
7      John    Physics     55
8      John  Chemistry    100
9     Arjun      Maths     90
10    Arjun    Physics     75
11    Arjun  Chemistry     70


In [11]:
# Create a new column with Marks  
# ranked in descending order 

df['Marks_Rank'] = df['Marks'].rank(ascending=0)
# Set index to newly created column  
df = df.set_index('Marks_Rank')
df

,Name,Subject,Marks
Marks_Rank,,,
4.5,Raj,Maths,80
2.5,Raj,Physics,90
6.5,Raj,Chemistry,75
9.5,Aravind,Maths,60
12.0,Aravind,Physics,40
9.5,Aravind,Chemistry,60
4.5,John,Maths,80
11.0,John,Physics,55
1.0,John,Chemistry,100


In [12]:
# Sort the DataFrame based on the index  
df = df.sort_index()
df

,Name,Subject,Marks
Marks_Rank,,,
1.0,John,Chemistry,100
2.5,Raj,Physics,90
2.5,Arjun,Maths,90
4.5,Raj,Maths,80
4.5,John,Maths,80
6.5,Raj,Chemistry,75
6.5,Arjun,Physics,75
8.0,Arjun,Chemistry,70
9.5,Aravind,Maths,60


Explanation:

Notice here that we have Raj and Arjun getting 90 marks each and hence they get ranked 2.5 (average of 2nd and 3rd rank i.e the two ranks they share). This can be seen for other marks in the table as well.